contracts/tf20/src/contract.rs

```diff
@@ -0,0 +1,541 @@
+        Some(admin) => {
+            if sender != admin {
+                return Err(ContractError::Unauthorized);
+            }
+            Ok(())
+        }
+    }
+}
+
+pub fn mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    recipient: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&recipient)?;
+
+    let denom = denom(deps.as_ref())?;
+    let coin = Coin {
+        denom,
+        amount: amount.clone().to_string(),
+    };
+
+    let force_transfer_msg = MsgMint {
+        sender: env.contract.address.into_string(),
+        amount: Some(coin),
+        mint_to_address: recipient.clone(),
+    };
+    let any_msg = AnyMsg {
+        type_url: String::from("/osmosis.tokenfactory.v1beta1.Msg/Mint"),
+        value: to_json_binary(&force_transfer_msg)?,
+    };
+
+    let res = Response::new()
+        .add_attribute("action", "mint")
+        .add_attribute("to", recipient)
+        .add_attribute("amount", amount)
+        .add_message(CosmosMsg::Any(any_msg));
+    Ok(res)
+}
+
+pub fn force_transfer(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    owner: String,
+    recipient: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&owner)?;
+    deps.api.addr_validate(&recipient)?;
+
+    _transfer(deps, env, owner, recipient, amount)
+}
+
+pub fn force_burn(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    owner: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&owner)?;
+    _burn(deps, env, owner, amount)
+}
+
+pub fn force_send(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    owner: String,
+    contract: String,
+    amount: Uint128,
+    msg: Binary,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender)?;
+
+    deps.api.addr_validate(&owner)?;
+    deps.api.addr_validate(&contract)?;
+    _send(deps, env, owner, contract, amount, msg)
+}
+
+pub fn update_contract_admin(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    new_admin: String,
+) -> Result<Response, ContractError> {
+    assert_admin(deps.as_ref(), info.sender.clone())?;
+    let old_admin = info.sender.into_string();
+
+    let admin = match new_admin.is_empty() {
+        true => None,
+        false => {
+            let addr = deps.api.addr_validate(&new_admin)?;
+            Some(addr)
+        }
+    };
+
+    let mut token_info = TOKEN_INFO.load(deps.storage)?;
+    token_info.admin = admin;
+    TOKEN_INFO.save(deps.storage, &token_info)?;
+    Ok(Response::new()
+        .add_attribute("action", "update_contract_admin")
+        .add_attribute("old_admin", old_admin)
+        .add_attribute("new_admin", new_admin))
+}
+
+pub fn update_tokenfactory_admin(
```

Comment:
should this also update the token info (internal contract state) to the new admin?

Reply:
This is basically a "Transfer the tokenfactory admin to someone else" move. The contract would no longer be an admin for the token (and would be pretty much useless). 

If you just wanted to transfer admin control of the contract to someone else, you would use that